File: src/main/java/com/morak/performancetracker/aop/persistence/StatementAdvice.java
package com.morak.performancetracker.aop.persistence;

import com.morak.performancetracker.context.Accumulator;
import org.aopalliance.intercept.MethodInterceptor;
import org.aopalliance.intercept.MethodInvocation;

public class StatementAdvice implements MethodInterceptor {

    private final QueryMonitor queryMonitor;
    private final Accumulator accumulator;

    public StatementAdvice(QueryMonitor queryMonitor, Accumulator accumulator) {
        this.queryMonitor = queryMonitor;
        this.accumulator = accumulator;
    }

    @Override
    public Object invoke(MethodInvocation invocation) throws Throwable {
        queryMonitor.start();
        Object returnValue = invocation.proceed();
        queryMonitor.end();
        accumulator.add(queryMonitor);
        return returnValue;
    }
}

File: src/main/kotlin/com/morak/performancetracker/aop/persistence/StatementAdvice.kt
package com.morak.performancetracker.aop.persistence

import com.morak.performancetracker.context.Accumulator
import org.aopalliance.intercept.MethodInterceptor
import org.aopalliance.intercept.MethodInvocation

class StatementAdvice(
    private val queryMonitor: QueryMonitor,
    private val accumulator: Accumulator,
) : MethodInterceptor {
    override fun invoke(invocation: MethodInvocation): Any? {
        queryMonitor.start()
        val returnValue = invocation.proceed()
        queryMonitor.end()
        accumulator.add(queryMonitor)
        return returnValue
    }
}
